adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,219 @@
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &bidResp); err != nil {
+
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: openrtb_ext.BidTypeBanner,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func setHeaders(raiHeaders *http.Header) {
+	raiHeaders.Set("Content-Type", "application/json;charset=utf-8")
+	raiHeaders.Set("Accept", "application/json")
+	raiHeaders.Add("X-Openrtb-Version", "2.5")
+}
+
+func setImp(request *openrtb2.BidRequest, isUrlSecure bool) (resImps []openrtb2.Imp, err error) {
+	for _, imp := range request.Imp {
+		var secure = int8(0)
+		raiExt, errImp := parseImpExt(&imp)
+		if errImp != nil {
+			return nil, errImp
+		}
+
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
+
+			if raiExt.Test {
+				request.Test = int8(1)
+			}
+
+			if raiExt.BidFloorCur != "" {
+				imp.BidFloorCur = raiExt.BidFloorCur
+			} else if imp.BidFloorCur == "" {
+				imp.BidFloorCur = "USD"
+			}
+		}
+		if isUrlSecure {
+			secure = int8(1)
+		}
+
+		imp.Secure = &secure
+
+		if imp.Banner.W == nil && imp.Banner.H == nil {
+			if len(imp.Banner.Format) == 0 {
+				err = &errortypes.BadInput{
+					Message: "request.Banner.Format is required",
+				}
+				return nil, err
+			}
+		}
+
+		resImps = append(resImps, imp)
+
+	}
+	return resImps, nil
+}
+
+func getIsUrlSecure(request *openrtb2.BidRequest) (isUrlSecure bool) {
+	if request.Site != nil {
+		if request.Site.Page != "" {
+			pageURL, err := url.Parse(request.Site.Page)
+			if err == nil {
+				if request.Site.Domain == "" {
+					request.Site.Domain = pageURL.Host
+				}
+				isUrlSecure = pageURL.Scheme == "https"
+			}
+		}
+	}
+	return
+}
+
+func validateDevice(request *openrtb2.BidRequest) (err error) {
+
+	if request.Device != nil && request.Device.IP == "" && request.Device.IPv6 == "" {
+		err = &errortypes.BadInput{
+			Message: "request.Device.IP is required",
+		}
+		return err
+	}
+	return err
+}
+
+func validateUser(request *openrtb2.BidRequest) (err error) {
+	if request.User != nil {
+		if request.User.Ext != nil {
+			var extUser openrtb_ext.ExtUser
+			if err := json.Unmarshal(request.User.Ext, &extUser); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
I noticed there is no test coverage for failure case of `validateUser`. Looking into why, I see this test is just validating the ability to unmarshal. You may remove this function from your adapter as this check is already performed by the core logic of PBS - specifically in the `validateUser` method of `endpoints/openrtb2/auction.go`.

Reply:
in next commit i have solved